### PR TITLE
Reuse cache revisited

### PR DESCRIPTION
**Pull request by totto82**
In the current code the cache is always invalidated, with this PR it is only invalidated when
1) the solution is updated
2) the time step fails

depend on OPM/opm-core#1117

Tested on SPE9, norne and Model 2.

### CODE REVIEW

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 1]
```diff
@@ -285,12 +281,11 @@ namespace Opm {
             std::vector<double> residual_norms;
             perfTimer.reset();
             perfTimer.start();
-            report.converged = getConvergence(timer, iteration,residual_norms);
+            // the step is not considered converged until at least minIter iterations is done
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: Is this change necessary? I'd like the report to state clearly that "yes, I converged" in this case, making it possible to tell the difference between the reasons for taking one more iteration (below min iters or not converged (== too large residuals in my mind)).

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 2]
```diff
@@ -338,14 +333,13 @@ namespace Opm {
                 // chopping of the update.
                 updateState(x,reservoir_state);
                 wellModel().updateWellState(xw, well_state);
+                // if the solution is updated the solution needs to be comunicated to ebos
+                // and the cachedIntensiveQuantities needs to be updated.
+                convertInput( iteration, reservoir_state, ebosSimulator_ );
+                ebosSimulator_.model().invalidateIntensiveQuantitiesCache(/*timeIdx=*/0);
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: Not sure how the timeIdx parameter is supposed to be used here? Should it not be the current timestep number?

Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: that's required because the model does not assume only the last "sub-solution" (in the Taylor series sense) of the time discretization to be variable. time index 0 just means "most recent sub-solution of the current time step".

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 1 above)
Comment (totto82) [inline opm/autodiff/BlackoilModelEbos.hpp]: If the residual is converged before any iterations, it will solve once (minIter = 1), updateState, invalidateCache, try to output (since report.converged is true) and fail.
With this change it will always linearize after the update, check  the residual again, if it is still converged it will output.
So  and alternative fix would be to call updateIntensiveQuantities if iter<minIter and report.converged. Do you want me to do that?

Comment (totto82) [inline opm/autodiff/BlackoilModelEbos.hpp]: Why do you need to know the reason for taking more iterations?
Adding a member "finished" in the simulatorReport would make it possible to distinguish the reason. It is a bit dangerous since most people would assume that converged == finished.

**Comment by andlaus** [comment]
what's the status of this? can it be merged? (I think it looks fine.)

**Comment by andlaus** [comment]
hm, I guess you forgot to add something. the travis error is

```
/home/travis/build/OPM/opm-simulators/opm/autodiff/BlackoilModelEbos.hpp:1466:24: error: ‘const class Opm::SimulatorTimerInterface’ has no member named ‘hasLastStepFailed’
```

**Comment by totto82** [comment]
Soon, I want to fix Atgeirr comment. It depends on OPM/opm-core#1117

**Comment by andlaus** [comment]
jenkins run norne_ebos opm-core=1117 please

**Comment by andlaus** [comment]
> It depends on OPM/opm-core#1117

oops, I guess I should have read the PR description more carefully instead of diving right into the code...

**Comment by atgeirr** [comment]
This must be updated as OPM/opm-core#1117 was changed before it was merged.

**Comment by atgeirr** [comment]
jenkins run norne please

**Comment by atgeirr** [comment]
I now think that perhaps a bug was added in OPM/opm-simulators#951, since there the test for minimum iterations in NonlinearSolver was changed from using `<=` to using '<'. Perhaps that has added to the confusion here. @andlaus, @totto82 any opinion? If you want to we could have a brief hangout to better discuss this.

**Comment by andlaus** [comment]
I can't remember changing this, but git has a much better memory than me. (that part code seems to be a bit error prone to me as well.) shall I open a PR?

**Comment by atgeirr** [comment]
> shall I open a PR?

Since this PR relates to the same logic, we could fix it here, if @totto82 thinks that is easy?

Otherwise: yes please.

**Comment by totto82** [comment]
@atgeirr Is this what you want?

**Comment by totto82** [comment]
jenkins run norne norne_ebos please

**Comment by totto82** [comment]
Model 2 finished with the same simulation time as ECL100 (2820s)!   

================    End of simulation     ===============

Total time taken (seconds):   2822.98
Solver time (seconds):        2775.2
 Assembly time (seconds):     920.09
 Linear solve time (seconds): 1482.98
 Update time (seconds):       27.7498
 Output write time (seconds): 50.4768
Overall Well Iterations:      584
Overall Linearizations:       2124
Overall Newton Iterations:    1783
Overall Linear Iterations:    45541

**Comment by atgeirr** [comment]
Thanks, I'll merge it now despite not agreeing fully about the treatment of the term "converged"... But that discussion can continue later.
